addons/mail/static/src/js/models/threads/thread.js

```diff
@@ -295,7 +295,7 @@ var Thread = AbstractThread.extend(ServicesMixin, {
     _generateEmojis: function (htmlString) {
         _.each(emojis, function (emoji) {
             _.each(emoji.sources, function (source) {
-                var escapedSource = String(source).replace(/([.*+?=^!:${}()|[\]/\\])/g, '\\$1');
+                var escapedSource = _.escape(String(source)).replace(/([.*+?=^!:${}()|[\]/\\])/g, '\\$1');
```

Comment:
Is it really needed to re-escape here? Won't this escape at every step of the loop, leading to not expected result?
If a change is needed, this should be tested as well.

Reply:
this is already tested, two assert in emoji tests breaks without this change.

in the lines:

https://github.com/odoo/odoo/blob/387c7abd2b4e82ba6078b867cc36b2b865302f7b/addons/mail/static/src/models/composer/composer.js#L225-L226

body is for example at first before `parseAndTransform`:

`:&#x27;)`

after parseAndTransform body is:

- before this PR  `:')`
- after this PR `:&#x27;)`

so if we don't break entities as done in this PR, in the `_generateEmojis` step you need to match with the escaped emoji source.

Reply:
Thank you for the details, this part of the code is such as mess, it's hard to be confident with any diff.